### DISCUSSION TITLE
Custom headers for API requests

### DISCUSSION THREAD

**Discussion by Dopingus** [BLeeEZ/amperfy#326] [Ideas]
Hello,

first of all thanks for creating Amperfy, it looks nice and feels modern. Very appreciated.

I would like to suggest the option to add custom headers to the API requests. This would enable me to put my Subsonic server behind a (Cloudflare) tunnel. Usually, the server URL would require authentication through some identity provider first (in my case it's Github), but by sending a specific Service Auth header, Amperfy would be able to bypass this authentication.

After seeing Symfonium (closed-source Subsonic client for Android) offering this option, I tried to setup my server with this additional layer of authentication and it worked just as expected.

While this might not increase security by a lot, I think it's still nice to be able to have an additional layer of authentication.

I tried my hand at building Amperfy locally on my Macbook, but for some reason it crashes on startup with the following exception:
```
Exception	NSException *	"Access to UINavigationBar's backdropGroupName ivar is prohibited. This is an application bug"	0x0000600000cb2070
```

Maybe I can work on this feature myself if I can get Amperfy to build.

Thanks & Regards

**Comment by BLeeEZ** [comment]
Hi,
which macOS version are you using? Are you already on 15.0?
I think this issues could be related to the dependence LNPopupController. Maybe you need to downgrade that dependency.
I assume that when you build for iOS this issue should not occur.

**Reply by Dopingus** [reply]
I am actually on 15.0 already. Thanks for the hint, I'll try to fiddle with that dependency.
The iOS build crashes with the same exception though, just later on.

**Reply by Dopingus** [reply]
Looks like LNPopupController had an update for this: https://github.com/LeoNatan/LNPopupController/issues/578
However, it looks like that breaks the MacCatalyst build. The iOS target still builds.

I'll continue to investigate.

**Reply by BLeeEZ** [reply]
I fixed that issue with 788b4e3dc880c3c81e70a63b8037c6856a018423. LNPopupController works with macOS in version 2.19.6. You should now be able to build without errors for macOS and iOS.

**Comment by Dopingus** [comment]
Sorry for the long delay. I finally got the time to setup XCode on a new machine and everything works fine now.
I created a draft for this feature: <https://github.com/BLeeEZ/amperfy/pull/396>

Do you use any formatter or linter? 

Due to not being super familiar with Swift I messed up formatting a lot, but using SwiftFormat for XCode also changes styles and order of imports etc.
Would you consider running some kind of formatter/linter and adding that to the README.md, so contributions do not change anything about the formatting/general style and can focus on actual code changes?

**Reply by BLeeEZ** [reply]
Hi @Dopingus,
I invested some time and introduced now a Code Formater [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) to apply the [Google Swift Style Guide](https://google.github.io/swift). I added a script to apply the style for all files and a build phase to Unit Tests. I added a section to the [Main Page](https://github.com/BLeeEZ/amperfy) specific for contributions and code styling.

Thank you for bringing this topic up.
